### PR TITLE
fix bug #185

### PR DESCRIPTION
**Pull request by berendeanicolae**
Fix inplace ProcessLastBlock.

### CODE REVIEW

File: modes.cpp [hunk 1]
```diff
@@ -207,9 +207,18 @@ void CBC_CTS_Encryption::ProcessLastBlock(byte *outString, const byte *inString,
...
 		m_cipher->ProcessBlock(m_register);
-		inString += BlockSize();
-		length -= BlockSize();
-		memcpy(outString+BlockSize(), m_register, length);
+		if (inString == outString)
+		{
+			length -= BlockSize();
+			memcpy(outString, inString+BlockSize(), length);
+			memcpy(outString+BlockSize(), m_register, length);
+		}
```
Comment (noloader) [inline modes.cpp]: When `inString == outString`, you have to use `memmove`, not `memcpy`.

Also see [Which one to use - memmove() or memcpy() - when buffers don't overlap?](http://stackoverflow.com/q/1960991/608639), [memcpy of overlapping buffers](http://stackoverflow.com/q/25629736/608639), [Glibc change exposing bugs](http://lwn.net/Articles/414467/) and [Strange sound on mp3 flash website](http://bugzilla.redhat.com/show_bug.cgi?id=638477) and

File: modes.cpp [hunk 2]
```diff
@@ -267,11 +276,23 @@ void CBC_CTS_Decryption::ProcessLastBlock(byte *outString, const byte *inString,
...
-		xorbuf(outString, m_temp, m_register, BlockSize());
+		if (inString == outString)
+		{
+			memcpy(outString, inString+BlockSize(), length);
+			memcpy(outString+BlockSize(), m_temp, length);
+			// decrypt next to last plaintext block
+			memcpy(m_temp, pn1, length);
+			m_cipher->ProcessBlock(m_temp);
+			xorbuf(outString, m_temp, m_register, BlockSize());
+		}
```
Comment (noloader) [inline modes.cpp]: Ditto.

**Comment by berendeanicolae** [comment]
Fixed. Thanks - I've never heard of memmove before

**Comment by noloader** [comment]
@berendeanicolae - I hope to start testing this soon. Its about number four on my list of things to do.

**Comment by berendeanicolae** [comment]
@noloader - I think that the operations should performed in a SecByteBlock (a secure block of memory) and not directly in the buffer. I'll make some time in time weekend to fix this and run the tests suite on a linux machine.

**Comment by berendeanicolae** [comment]
@noloader - I've modified ProcessLastBlock within CBC_CTS so that the stolenIV implementation is separated from the rest and now all the processing (encryption, xor, etc.) is done in a SecByteBlock.

I've run the tests from the pr ([pastebin](http://pastebin.com/3ubfLwGS)) on win x64 and they worked. Then I've run cryptest.sh on arch linux x64 and all the tests worked.

**Comment by noloader** [comment]
@berendeanicolae,

I'm not sure what to do with this at the moment. I have not had the time to investigate it further.

Before anything happens, I'd like to get some test cases added. Test cases are a great first step. I think they are mandatory in this case to help us understand what's going on and to test the fix is a complete remdiation on all platforms.

I'm going to close this at the moment. I will probably ping you in the future to coordinate when I have the time to deep dive the issue.

Also, there could be a lurking underlying high-level policy/procedure issue. That issue is the in-place encryption and decryption. We may not have policy and procedures in place to deal with it. I hope that's not the case because the efficiencies gained with in-place transformations should not be taken for granted.

**Comment by berendeanicolae** [comment]
@noloader 

So, would you like CBC_CTS to be patched to support in-place transformation?

It's true that in the documentation it's not specified that CBC_CTS supports in-place transformation but, as CBC_CTS is derived from CBC which supports in-place transformation CBC_CTS should also support it.

You said that you'd like some test cases to be added. I saw the tests are implemented in validat1.cpp. Do you want me to create some tests, or rather put the issue on hold for now?

**Pull request closed without merging**